csrc/tensor_metadata.cpp

```diff
@@ -323,30 +323,64 @@ inferAndValidateAllocationSizesAndStrides(
   return {std::move(sizes), std::move(strides)};
 }
 
+namespace {
+std::pair<std::vector<int64_t>, std::vector<int64_t>> unshardedSizesAndStrides(
+    TensorView* tv,
+    c10::IntArrayRef sizes,
+    c10::IntArrayRef strides) {
+  std::vector<int64_t> unsharded_sizes(sizes.size());
+  std::vector<int64_t> unsharded_strides(strides.size());
+  for (const auto i : c10::irange(sizes.size())) {
+    IterDomain* id = tv->getLogicalDomain()[i];
+    if (id->isDeviceDim()) {
+      unsharded_sizes[i] = tv->getDeviceMesh().size(id->getParallelType());
+      // This probably doesn't matter in practice unless a kernel accidentally
+      // tries to access the data on another rank. To be safe, set the stride
+      // to zero, analogous to an expanded broadcast dimension.
+      unsharded_strides[i] = 0;
```

Comment:
Using 0 is a good idea. I think it makes it more explicit the axis isn't allocated.